pallets/xcmp-queue/src/lib.rs

```diff
@@ -25,29 +25,145 @@
 
 #![cfg_attr(not(feature = "std"), no_std)]
 
-use sp_std::{prelude::*, convert::TryFrom};
-use rand_chacha::{rand_core::{RngCore, SeedableRng}, ChaChaRng};
 use codec::{Decode, Encode};
-use sp_runtime::{RuntimeDebug, traits::Hash};
-use frame_support::{decl_error, decl_event, decl_module, decl_storage, dispatch::Weight};
-use xcm::{
-	VersionedXcm, v0::{
-		Error as XcmError, ExecuteXcm, Junction, MultiLocation, SendXcm, Outcome, Xcm,
-	},
-};
 use cumulus_primitives_core::{
-	XcmpMessageHandler, ParaId, XcmpMessageSource, ChannelStatus, MessageSendError, GetChannelInfo,
-	relay_chain::BlockNumber as RelayBlockNumber,
+	relay_chain::BlockNumber as RelayBlockNumber, ChannelStatus, GetChannelInfo, MessageSendError,
+	ParaId, XcmpMessageHandler, XcmpMessageSource,
+};
+use frame_support::weights::Weight;
+use rand_chacha::{
+	rand_core::{RngCore, SeedableRng},
+	ChaChaRng,
 };
+use sp_runtime::{traits::Hash, RuntimeDebug};
+use sp_std::{convert::TryFrom, prelude::*};
+use xcm::{
+	v0::{Error as XcmError, ExecuteXcm, Junction, MultiLocation, Outcome, SendXcm, Xcm},
+	VersionedXcm,
+};
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Something to execute an XCM message. We need this to service the XCMoXCMP queue.
+		type XcmExecutor: ExecuteXcm<Self::Call>;
+
+		/// Information on the avaialble XCMP channels.
+		type ChannelInfo: GetChannelInfo;
+	}
 
-pub trait Config: frame_system::Config {
-	type Event: From<Event<Self>> + Into<<Self as frame_system::Config>::Event>;
+	impl Default for QueueConfigData {
+		fn default() -> Self {
+			Self {
+				suspend_threshold: 2,
+				drop_threshold: 5,
+				resume_threshold: 1,
+				threshold_weight: 100_000,
+				weight_restrict_decay: 2,
+			}
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_idle(_now: T::BlockNumber, max_weight: Weight) -> Weight {
+			// on_idle processes additional messages with any remaining block weight.
+			Self::service_xcmp_queue(max_weight)
+		}
+	}
 
-	/// Something to execute an XCM message. We need this to service the XCMoXCMP queue.
-	type XcmExecutor: ExecuteXcm<Self::Call>;
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
 
-	/// Information on the avaialble XCMP channels.
-	type ChannelInfo: GetChannelInfo;
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	#[pallet::metadata(T::Hash = "Hash")]
+	pub enum Event<T: Config> {
+		/// Some XCM was executed ok.
+		Success(Option<T::Hash>),
```

Comment:
I think the metadata will contains the type `"Option<T::Hash>` and not `Option<Hash>` as wanted.

Reply:
Wouldn't the attribute above rewrite it to `Option<Hash>`?

Reply:
I don't think so, did you try by comparing metadata before and after the implementation ?
In order to solve it the metadata attribute should contain: `Option<T::Hash> = "Option<Hash>"`

I know we could improve the macro maybe, but I think the metadata will be changed to use scale-info in a not too distant future.

Reply:
I didn't check specifically on this pallet, but while I was upgrading some other pallets in substrate, type parameters do get changed by the metadata attribute. Will build and check on this pallet as well.

Reply:
You're right, it doesn't change it, let me push a commit to fix this.